### PR TITLE
Spike/dockerizing

### PR DESCRIPTION
**Pull request by bogdanadrianmarc**
This PR aims to dockerize the app. The Dockerfile is using the full `ruby` package rather than the `alpine` version because `libv8` (required by `therubyracer` gem) doesn't support the `alpine` version

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# Set working dir and copy app
+WORKDIR /usr/src/app
+COPY . .
+
+# Prerequisites for gems install
+RUN apt-get install tzdata \
+                    git
+
+# Install bundler and gems
+RUN gem install bundler:2.1.4
```
Comment (ijdickinson) [inline Dockerfile:16]: I'm not sure we should fix the `bundler` version in the Dockerfile like this. If it needs to be specified (and I'm not sure that it does), I think it should be an arg

**Review by ijdickinson — CHANGES_REQUESTED** [review]
Still to do:

- JSON log output to stdout
- serving static files
- secret key base
- specifying the location of the API

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Params
+ARG RAILS_ENV="production"
+
+# Defining ruby version
+FROM ruby:2.6.6
```
Comment (mika018) [inline Dockerfile:5]: I would parametarise the ruby version so that we have control over it. Try this:
```
ARG RUBY_VERSION=2.6.6 
FROM ruby:$RUBY_VERSION
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+# Set working dir and copy app
+WORKDIR /usr/src/app
+COPY . .
+
+# Prerequisites for gems install
+RUN apt-get install tzdata \
+                    git
+
+# Install bundler and gems
+RUN gem install bundler
```
Comment (mika018) [inline Dockerfile:16]: Similarly I would parametarise the bundler version. Add `ARG BUNDLER_VERSION=2.2.16` (use an appropriate default) to the top of the file and then rewrite this line as `RUN gem install bundler:$BUNDLER_VERSION`

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+# Install bundler and gems
+RUN gem install bundler:$BUNDLER_VERSION
+RUN bundle install
+
+# Params
+ARG RAILS_ENV="production"
+ARG RAILS_SERVE_STATIC_FILES="true"
+ARG RELATIVE_URL_ROOT="/app/ppd"
+ARG API_SERVICE_URL="http://localhost:8080"
```
Comment (ijdickinson) [inline Dockerfile:23]: I'm confused by the API_SERVICE_URL defaulting to `localhost:8080`. That will only be useful if a developer happens to be running a development copy of Sapi-NT locally. I know  you are doing that at the moment, but that's not how we will usually work. I think it would be better to make the default URL be one that will "just work" if someone needs to make a quick change, and otherwise if we happen to be running Sapi-NT locally then we can pass in `localhost:8080` as an arg.

File: entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+# Set the environment
+if [ -z "$RAILS_ENV" ]
+then
+  export RAILS_ENV=production
+fi
+
+# Handle secrets based on env
+if [ "$RAILS_ENV" = "production" ]
```
Comment (ijdickinson) [inline entrypoint.sh:10]: Could this be an `and` condition, rather than nested `if`s?

**Review by ijdickinson — APPROVED** [review]
Looks good, I think we're nearly there. I did have a question about which default URL to bake into the script.

I don't think we have enough of the system up and running to run the `hmlr-apps-feature-tests` in the normal way yet, but you could manually try out the tests from https://github.com/epimorphics/hmlr-apps-feature-tests/blob/master/specs/feature/ppd_test.rb to see if you get plausible looking results

File: config/environments/production.rb [hunk 6]
```diff
@@ -83,8 +85,12 @@
...
 
   # Specify that we're not at the root
-  config.action_controller.relative_url_root = '/app/ppd'
-  config.relative_url_root = '/app/ppd'
+  config.action_controller.relative_url_root = ENV['RELATIVE_URL_ROOT'] || '/app/ppd'
+  config.relative_url_root = ENV['RELATIVE_URL_ROOT'] || '/app/ppd'
+
+  # API location can be specified in the environment
+  # But defaults to the dev service
+  config.api_service_url = ENV['API_SERVICE_URL'] || 'http://localhost:8080'
```
Comment (mika018) [inline config/environments/production.rb:93]: Similarly here as in @ijdickinson mentioned above. We should probably change the default here to be a meaningful location where you would usually find the API service. I'm not exactly sure what is the right location though.

File: config/secrets.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+# You can use `rake secret` to generate a secure secret key.
+
+# Make sure the secrets in this file are kept private
+# if you're sharing your code publicly.
+
+development:
+  secret_key_base: 436cc31267e1dfe7bffcf805e537eefb589cb47057a3e572ba9f46f71bd7b4b2bf6fe6653835ea3541aa2c0ba57b3de3d33b122e7a9923e8e6019cb7a9852a23
+
+test:
+  secret_key_base: 05842753351a9f1603a2e8a703c2c793c01143c216e5177b17bf8a937319926bc8a8ac2e4e828f0c60b07d8fac8bfe27e8dbfcb95e02ee229a9e976da85cbe22
```
Comment (mika018) [inline config/secrets.yml:17]: Do we want this file commited? It is usually a bad practice to commit keys even if they are just used for development. I can see that we are using similar approach in reg products but this file isn't present in that repo, not sure if that approach can be used here, just a suggestion.

Comment (ijdickinson) [reply config/secrets.yml:17]: This is the default that happens when you create a new Rails project with `rails new`. I think the thinking is: projects are only run for real with `RAILS_ENV=production`, so the dev and test keys (which are randomly generated by every `rails new`) are benign to commit to source control. And if they weren't, every time we ran the test suite, locally or in CI, we'd need to capture the output of `rails secret` in an env var.

Tl;dr - good question to ask, but I think it's OK

File: config/environments/production.rb (hunk 6 above)
Comment (ijdickinson) [reply config/environments/production.rb:93]: To me, it depends if we're keeping a `dev` service around. If we are, that should be the default. If not, probably the `production` API, since pre-prod is protected by ssh firewall-ery

File: Gemfile [hunk 8]
```diff
@@ -39,7 +41,9 @@ group :development do
   gem 'stackprof' # ruby 2.1+ only
 end
 
-gem 'data_services_api', git: 'https://github.com/epimorphics/ds-api-ruby.git'
+# rubocop:disable Layout/LineLength
```
Comment (ijdickinson) [inline Gemfile:44]: I don't care about Rubocop's opinion of the `Gemfile`, so I think rather than toggle Cops on and off in this file, we should just add an `Exclude` line to `.rubocop.yml`

File: config/environments/development.rb [hunk 9]
```diff
@@ -32,7 +32,7 @@
   config.log_tags = %i[subdomain request_id request_method]
   # config.logger = ActiveSupport::TaggedLogging.new(SyslogLogger.new)
 
-  config.api_service_url = ENV['API_SERVICE_URL'] || 'https://lr-ppd-dev-pres.epimorphics.net'
+  config.api_service_url = ENV['API_SERVICE_URL'] || 'http://localhost:8080'
```
Comment (ijdickinson) [inline config/environments/development.rb:35]: I thought we were going to make the default URL point to the new dev service?

File: config/environments/production.rb [hunk 10]
```diff
@@ -83,8 +85,12 @@
   config.log_formatter = ::Logger::Formatter.new
 
   # Specify that we're not at the root
-  config.action_controller.relative_url_root = '/app/ppd'
-  config.relative_url_root = '/app/ppd'
+  config.action_controller.relative_url_root = ENV['RELATIVE_URL_ROOT'] || '/app/ppd'
```
Comment (ijdickinson) [inline config/environments/production.rb:88]: Looking at this, I'm wondering now if it would make more sense to either make the default `/`, or to assume that the `RELATIVE_URL_ROOT` will always be passed in as a param when running in production mode? Having a baked-in default of `/app/ppd` looks out of place here now

File: config/environments/test.rb [hunk 11]
```diff
@@ -36,6 +36,10 @@
   # Print deprecation notices to the stderr.
   config.active_support.deprecation = :stderr
 
+  # API location can be specified in the environment
+  # But defaults to the dev service
+  config.api_service_url = ENV['API_SERVICE_URL'] || 'http://localhost:8080'
```
Comment (ijdickinson) [inline config/environments/test.rb:41]: Again, I'm not sure the default should be `localhost`

**Review by ijdickinson — COMMENTED** [review]
Looks like we're nearly there. Just a few small tweaks needed, I think

File: config/environments/test.rb [hunk 12]
```diff
@@ -36,6 +36,10 @@
   # Print deprecation notices to the stderr.
   config.active_support.deprecation = :stderr
 
+  # API location can be specified in the environment
+  # But defaults to the dev service
+  config.api_service_url = ENV['API_SERVICE_URL'] || 'https://lr-ppd-dev-pres.epimorphics.net'
```
Comment (ijdickinson) [inline config/environments/test.rb:41]: Per my message on Slack, there's no point defaulting this URL to a service that is not running. If we can't point to the new service yet, we should leave the default out

**Review by ijdickinson — COMMENTED** [review]
Nearly there. But we should remove the reference to a `dev` service that is (a) switched off, and (b) even if it wasn't switched off it would be using the old-style DsAPI not Sapi.
